runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -2625,3 +2625,164 @@ J9::TransformUtil::refineMethodHandleLinkTo(TR::Compilation* comp, TR::TreeTop*
+   //     the dataAddr pointer field then add it to the offset.
+   // 3.) The type of the object at dest is unknown at compile time (type is `java/lang/Object`). In this scenario,
+   //     a runtime null check and arrayCHK must be generated to determine whether it needs to be handled such as
+   //     case A or case B.
+
+   TR::Node *src        = arraycopyNode->getChild(1);
+   TR::Node *dest       = arraycopyNode->getChild(3);
+
+   // Check src/dest type at compile time
+   int srcSigLen, destSigLen;
+   const char *srcObjTypeSig = src->getSymbolReference() ? src->getSymbolReference()->getTypeSignature(srcSigLen) : 0;
+   const char *destObjTypeSig = dest->getSymbolReference() ? dest->getSymbolReference()->getTypeSignature(destSigLen) : 0;
+   // Case 3
+   bool srcArrayCheckNeeded = srcObjTypeSig == NULL ||
+                                 strncmp(srcObjTypeSig, "Ljava/lang/Object;", strlen("Ljava/lang/Object;")) == 0 ||
+                                 strncmp(srcObjTypeSig, "Ljava/lang/Cloneable;", strlen("Ljava/lang/Cloneable;")) == 0 ||
+                                 strncmp(srcObjTypeSig, "Ljava/io/Serializable;", strlen("Ljava/io/Serializable;")) == 0;
+   bool destArrayCheckNeeded = destObjTypeSig == NULL ||
+                                 strncmp(destObjTypeSig, "Ljava/lang/Object;", strlen("Ljava/lang/Object;")) == 0 ||
+                                 strncmp(destObjTypeSig, "Ljava/lang/Cloneable;", strlen("Ljava/lang/Cloneable;")) == 0 ||
+                                 strncmp(destObjTypeSig, "Ljava/io/Serializable;", strlen("Ljava/io/Serializable;")) == 0;
+   // Case 2 & 3
+   bool srcAdjustmentNeeded = srcArrayCheckNeeded || srcObjTypeSig[0] == '[';
+   bool destAdjustmentNeeded = destArrayCheckNeeded || destObjTypeSig[0] == '[';
+
+   TR::TransformUtil::separateNullCheck(comp, arrayCopyTT);
+
+   if (!(srcAdjustmentNeeded || destAdjustmentNeeded))
+      {
+      TR::TransformUtil::convertUnsafeCopyMemoryCallToArrayCopyWithSymRefLoad(comp, arrayCopyTT, NULL, NULL);
+      return;
+      }
+
+   // Anchor nodes
+   for (int32_t i=1; i < arraycopyNode->getNumChildren(); i++)
+      {
+      TR::Node* childNode = arraycopyNode->getChild(i);
+      if ( !(childNode->getOpCode().isLoadConst() || childNode->getSymbolReference()->getSymbol()->isAutoOrParm()) )
+         arrayCopyTT->insertBefore(TR::TreeTop::create(comp, TR::Node::create(TR::treetop, 1, arraycopyNode->getChild(i))));
+      }
+
+   TR::CFG *cfg = comp->getFlowGraph();
+   TR::Block *currentBlock = arrayCopyTT->getEnclosingBlock();
+   TR::Block *callBlock = currentBlock->split(arrayCopyTT, cfg, true);
+   TR::Block *nextBlock = callBlock->split(arrayCopyTT->getNextTreeTop(), cfg, true);
+
+   TR::SymbolReference *adjustSrcTempRef = NULL;
+   TR::SymbolReference *adjustDestTempRef = NULL;
+   if (srcAdjustmentNeeded)
+      {
+      src->createStoresForVar(adjustSrcTempRef, currentBlock->getExit());
```

Comment:
I think this is still creating a collected temp to store into, and similarly for `dest`

Reply:
We still need this temp as either arrayBase or dataAddr, then load it in the arrayCopy call.
Starts with the arrayBase, overridden with the dataAddr if needed, then loaded in the arrayCopy call.

Reply:
Yes, the temp should be non-collected (https://github.com/eclipse-openj9/openj9/pull/19634#discussion_r1649458628)